easybuild/main.py

```diff
@@ -213,162 +214,100 @@ def clean_exit(logfile, tmpdir, testing, silent=False):
     sys.exit(0)
 
 
-def main(args=None, logfile=None, do_build=None, testing=False, modtool=None):
+def process_easystack(easystack_path, args, logfile, testing, init_session_state, do_build):
     """
-    Main function: parse command line options, and act accordingly.
-    :param args: command line arguments to use
+    Process an EasyStack file. That means, parsing, looping over all items in the EasyStack file
+    building (where requested) the individual items, etc
+
+    :param easystack_path: path to EasyStack file to be processed
+    :param args: original command line arguments as received by main()
     :param logfile: log file to use
-    :param do_build: whether or not to actually perform the build
     :param testing: enable testing mode
+    :param init_session_state: initial session state, to use in test reports
+    :param do_build: whether or not to actually perform the build
     """
-
-    register_lock_cleanup_signal_handlers()
-
-    # if $CDPATH is set, unset it, it'll only cause trouble...
-    # see https://github.com/easybuilders/easybuild-framework/issues/2944
-    if 'CDPATH' in os.environ:
-        del os.environ['CDPATH']
-
-    # When EB is run via `exec` the special bash variable $_ is not set
-    # So emulate this here to allow (module) scripts depending on that to work
-    if '_' not in os.environ:
-        os.environ['_'] = sys.executable
-
-    # purposely session state very early, to avoid modules loaded by EasyBuild meddling in
-    init_session_state = session_state()
-
-    eb_go, cfg_settings = set_up_configuration(args=args, logfile=logfile, testing=testing)
-    options, orig_paths = eb_go.options, eb_go.args
-
-    global _log
-    (build_specs, _log, logfile, robot_path, search_query, eb_tmpdir, try_to_generate,
-     from_pr_list, tweaked_ecs_paths) = cfg_settings
-
-    # load hook implementations (if any)
-    hooks = load_hooks(options.hooks)
-
-    run_hook(START, hooks)
-
-    if modtool is None:
+    orig_paths, opts_per_ec = parse_easystack(easystack_path)
+
+    _log.debug("Start build loop over items in the EasyStack file: %s" % orig_paths)
+    # TODO: insert fast loop that validates if all command line options are valid. If there are errors in options,
+    # we want to know early on, and this loop potentially builds a lot of packages and could take very long
+    # for path in orig_paths:
+    #     validate_command_opts(args, opts_per_ec[path])
+
+    # Loop over each item in the EasyStack file, each time updating the config
+    # This is because each item in an EasyStack file can have options associated with it
+    do_cleanup = True
+    for path in orig_paths:
+        _log.debug("Starting build for %s" % path)
+        # Whipe easyconfig caches
+        easyconfig._easyconfigs_cache.clear()
```

Comment:
@boegel I'm not sure this is actually needed. I created the following test:
```
cat test_clear_easyconfig_cache.yaml
robot: true
easyconfigs:
  - GCC-11.3.0.eb
  - binutils-2.38-GCCcore-11.3.0.eb:
      options: {
        'from-pr': '16386',
      }
  - GCC-11.3.0.eb
```
My assumption was that, without clearing the cache, the same EasyConfig file would be used every time for `binutils-2.38-GCCcore-11.3.0.eb`. But, even if I comment these two `clear()` lines out, I get

```
Dry run: printing build status of easyconfigs and dependencies
CFGS=/sw/noarch/Debian10/2022/software/EasyBuild/4.6.1/easybuild/easyconfigs
 * [x] $CFGS/m/M4/M4-1.4.19.eb (module: M4/1.4.19)
 * [x] $CFGS/b/Bison/Bison-3.8.2.eb (module: Bison/3.8.2)
 * [x] $CFGS/f/flex/flex-2.6.4.eb (module: flex/2.6.4)
 * [x] $CFGS/z/zlib/zlib-1.2.12.eb (module: zlib/1.2.12)
 * [x] $CFGS/b/binutils/binutils-2.38.eb (module: binutils/2.38)
 * [x] $CFGS/g/GCCcore/GCCcore-11.3.0.eb (module: GCCcore/11.3.0)
 * [x] $CFGS/z/zlib/zlib-1.2.12-GCCcore-11.3.0.eb (module: zlib/1.2.12-GCCcore-11.3.0)
 * [x] $CFGS/h/help2man/help2man-1.49.2-GCCcore-11.3.0.eb (module: help2man/1.49.2-GCCcore-11.3.0)
 * [x] $CFGS/m/M4/M4-1.4.19-GCCcore-11.3.0.eb (module: M4/1.4.19-GCCcore-11.3.0)
 * [x] $CFGS/b/Bison/Bison-3.8.2-GCCcore-11.3.0.eb (module: Bison/3.8.2-GCCcore-11.3.0)
 * [x] $CFGS/f/flex/flex-2.6.4-GCCcore-11.3.0.eb (module: flex/2.6.4-GCCcore-11.3.0)
 * [x] $CFGS/b/binutils/binutils-2.38-GCCcore-11.3.0.eb (module: binutils/2.38-GCCcore-11.3.0)
 * [x] $CFGS/g/GCC/GCC-11.3.0.eb (module: GCC/11.3.0)
== Temporary log file in case of crash /scratch/casparl/eb-l8wbjlkf/easybuild-164xa9n9.log
Dry run: printing build status of easyconfigs and dependencies
 * [x] /sw/noarch/Debian10/2022/software/EasyBuild/4.6.1/easybuild/easyconfigs/m/M4/M4-1.4.19.eb (module: M4/1.4.19)
 * [x] /sw/noarch/Debian10/2022/software/EasyBuild/4.6.1/easybuild/easyconfigs/b/Bison/Bison-3.8.2.eb (module: Bison/3.8.2)
 * [x] /sw/noarch/Debian10/2022/software/EasyBuild/4.6.1/easybuild/easyconfigs/f/flex/flex-2.6.4.eb (module: flex/2.6.4)
 * [x] /sw/noarch/Debian10/2022/software/EasyBuild/4.6.1/easybuild/easyconfigs/z/zlib/zlib-1.2.12.eb (module: zlib/1.2.12)
 * [x] /sw/noarch/Debian10/2022/software/EasyBuild/4.6.1/easybuild/easyconfigs/b/binutils/binutils-2.38.eb (module: binutils/2.38)
 * [x] /sw/noarch/Debian10/2022/software/EasyBuild/4.6.1/easybuild/easyconfigs/g/GCCcore/GCCcore-11.3.0.eb (module: GCCcore/11.3.0)
 * [x] /sw/noarch/Debian10/2022/software/EasyBuild/4.6.1/easybuild/easyconfigs/h/help2man/help2man-1.49.2-GCCcore-11.3.0.eb (module: help2man/1.49.2-GCCcore-11.3.0)
 * [x] /sw/noarch/Debian10/2022/software/EasyBuild/4.6.1/easybuild/easyconfigs/m/M4/M4-1.4.19-GCCcore-11.3.0.eb (module: M4/1.4.19-GCCcore-11.3.0)
 * [x] /sw/noarch/Debian10/2022/software/EasyBuild/4.6.1/easybuild/easyconfigs/z/zlib/zlib-1.2.12-GCCcore-11.3.0.eb (module: zlib/1.2.12-GCCcore-11.3.0)
 * [x] /sw/noarch/Debian10/2022/software/EasyBuild/4.6.1/easybuild/easyconfigs/b/Bison/Bison-3.8.2-GCCcore-11.3.0.eb (module: Bison/3.8.2-GCCcore-11.3.0)
 * [x] /sw/noarch/Debian10/2022/software/EasyBuild/4.6.1/easybuild/easyconfigs/f/flex/flex-2.6.4-GCCcore-11.3.0.eb (module: flex/2.6.4-GCCcore-11.3.0)
 * [x] /scratch/casparl/eb-blcb26tg/files_pr16386/b/binutils/binutils-2.38-GCCcore-11.3.0.eb (module: binutils/2.38-GCCcore-11.3.0)
== Temporary log file in case of crash /scratch/casparl/eb-l8wbjlkf/easybuild-164xa9n9.log
== found valid index for /sw/noarch/Debian10/2022/software/EasyBuild/4.6.1/easybuild/easyconfigs, so using it...
Dry run: printing build status of easyconfigs and dependencies
CFGS=/sw/noarch/Debian10/2022/software/EasyBuild/4.6.1/easybuild/easyconfigs
 * [x] $CFGS/m/M4/M4-1.4.19.eb (module: M4/1.4.19)
 * [x] $CFGS/b/Bison/Bison-3.8.2.eb (module: Bison/3.8.2)
 * [x] $CFGS/f/flex/flex-2.6.4.eb (module: flex/2.6.4)
 * [x] $CFGS/z/zlib/zlib-1.2.12.eb (module: zlib/1.2.12)
 * [x] $CFGS/b/binutils/binutils-2.38.eb (module: binutils/2.38)
 * [x] $CFGS/g/GCCcore/GCCcore-11.3.0.eb (module: GCCcore/11.3.0)
 * [x] $CFGS/z/zlib/zlib-1.2.12-GCCcore-11.3.0.eb (module: zlib/1.2.12-GCCcore-11.3.0)
 * [x] $CFGS/h/help2man/help2man-1.49.2-GCCcore-11.3.0.eb (module: help2man/1.49.2-GCCcore-11.3.0)
 * [x] $CFGS/m/M4/M4-1.4.19-GCCcore-11.3.0.eb (module: M4/1.4.19-GCCcore-11.3.0)
 * [x] $CFGS/b/Bison/Bison-3.8.2-GCCcore-11.3.0.eb (module: Bison/3.8.2-GCCcore-11.3.0)
 * [x] $CFGS/f/flex/flex-2.6.4-GCCcore-11.3.0.eb (module: flex/2.6.4-GCCcore-11.3.0)
 * [x] $CFGS/b/binutils/binutils-2.38-GCCcore-11.3.0.eb (module: binutils/2.38-GCCcore-11.3.0)
 * [x] $CFGS/g/GCC/GCC-11.3.0.eb (module: GCC/11.3.0)
```
I did not expect that to work with the EasyConfig cache not being whiped... Is it possible that something else (e.g. the `set_up_configuration`) is already clearing the cache / creating a new clean cache? In that case, these two lines could be removed.

Reply:
The problem here was that you were listed the same easyconfig twice in the same easystack. That didn't work due to how we were looping over easystack items, which we changed in 63a879d